### ISSUE TITLE
Add a ShufflingDataSetIterator

### ISSUE THREAD

**Issue by wmeddie** [deeplearning4j/nd4j#1837]
#### Issue Description

Since the DataSetIterator API allows for DataSets too large to fit in memory we should forego true uniform shuffling with a sampling method like Fisher-Yates.  Users should be able to specify how uniform the shuffling is by specifying how large the buffer can be.

**Comment by raver119** [comment]
This thing should implemented with caution & respect to Workspaces.

Ideally, if buffer is instantiated once, _not in iterator next() mehthod_ and reused during iterator lifetime.

**Comment by wmeddie** [comment]
Oh.  That complicates things.

**Comment by raver119** [comment]
Yes, slightly, but not that much really :)

**Comment by huitseeker** [comment]
See also https://github.com/deeplearning4j/nd4j/issues/1521

**Comment by AlexDBlack** [comment]
My more recent thinking on the subject is that earlier in the pipeline is more appropriate for shuffling - for example, DataVec's (new) MapFileRecordReader can support (proper, not just local/pseudo-random) shuffling easily, and has none of the memory overhead of a shuffling DataSetIterator. Images etc. can also support this.

@wmeddie I'd be curious to hear of use cases where you couldn't shuffle at the RecordReader level though, and would want to shuffle at the DataSetIterator level.

**Comment by wmeddie** [comment]
@AlexDBlack I did shuffle the data once and then serialized the minibatches to disk for performance reasons.  (about 1 GB of datasets). I could create about 10 different shuffles and export them separately but it seems like a waste to me since it's the same data.

There are also some datasets (like MNISTor CIFAR) that you don't have an easy way to access with RecordReaders, but you still might want to shuffle them to give the network a better chance at not getting stuck.  

I'm for proper shuffling for making train/test/validation sets or k-folds, but the main selling point of ShufflingDataSetIterator is that it's just a simple way of putting a just a bit more stochasticity into the gradient descent.
